src/target/riscv/riscv-013.c

```diff
@@ -1272,193 +1286,224 @@ static bool has_sufficient_progbuf(struct target *target, unsigned size)
-				if (riscv_program_insert(&program,
-							fmv_d_x(number - GDB_REGNO_FPR0, S0)) != ERROR_OK)
-					return ERROR_FAIL;
-			} else {
-				if (riscv_program_insert(&program,
-							fmv_w_x(number - GDB_REGNO_FPR0, S0)) != ERROR_OK)
-					return ERROR_FAIL;
-			}
-		} else if (number == GDB_REGNO_VTYPE) {
-			if (riscv_save_register(target, GDB_REGNO_S1) != ERROR_OK)
-				return ERROR_FAIL;
-			if (riscv_program_insert(&program, csrr(S1, CSR_VL)) != ERROR_OK)
-				return ERROR_FAIL;
-			if (riscv_program_insert(&program, vsetvl(ZERO, S1, S0)) != ERROR_OK)
-				return ERROR_FAIL;
-		} else if (number == GDB_REGNO_VL) {
-			/* "The XLEN-bit-wide read-only vl CSR can only be updated by the
-			 * vsetvli and vsetvl instructions, and the fault-only-rst vector
-			 * load instruction variants." */
-			if (riscv_save_register(target, GDB_REGNO_S1) != ERROR_OK)
-				return ERROR_FAIL;
-			if (riscv_program_insert(&program, csrr(S1, CSR_VTYPE)) != ERROR_OK)
-				return ERROR_FAIL;
-			if (riscv_program_insert(&program, vsetvl(ZERO, S0, S1)) != ERROR_OK)
-				return ERROR_FAIL;
-		} else if (number >= GDB_REGNO_CSR0 && number <= GDB_REGNO_CSR4095) {
-			if (riscv_program_csrw(&program, S0, number) != ERROR_OK)
+		if (riscv_supports_extension(target, 'D') && riscv_xlen(target) < 64) {
+			/* There are no instructions to move all the bits from a
+			 * register, so we need to use some scratch RAM.
+			 */
+			if (riscv_program_insert(&program, fsd(freg, S0, 0)) != ERROR_OK)
 				return ERROR_FAIL;
-		} else {
-			LOG_ERROR("Unsupported register (enum gdb_regno)(%d)", number);
-			return ERROR_FAIL;
+			return internal_register_read64_progbuf_scratch(target, &program,
+					value);
 		}
-		if (register_write_direct(target, GDB_REGNO_S0, value) != ERROR_OK)
+		if (riscv_program_insert(&program,
+					riscv_supports_extension(target, 'D') ?
+					fmv_x_d(S0, freg) : fmv_x_w(S0, freg)) != ERROR_OK)
 			return ERROR_FAIL;
+	} else if (number >= GDB_REGNO_CSR0 && number <= GDB_REGNO_CSR4095) {
+		if (riscv_program_csrr(&program, S0, number) != ERROR_OK)
+			return ERROR_FAIL;
+	} else {
+		LOG_ERROR("Unsupported register: %s", gdb_regno_name(number));
+		return ERROR_FAIL;
 	}
 
-	int exec_out = riscv_program_exec(&program, target);
-	/* Don't message on error. Probably the register doesn't exist. */
-	if (exec_out == ERROR_OK && target->reg_cache) {
-		struct reg *reg = &target->reg_cache->reg_list[number];
-		buf_set_u64(reg->value, 0, reg->size, value);
-	}
+	if (riscv_program_exec(&program, target) != ERROR_OK)
+		return ERROR_FAIL;
 
-	if (use_scratch)
-		scratch_release(target, &scratch);
+	return register_read_direct(target, value, GDB_REGNO_S0) != ERROR_OK;
+}
 
-	if (cleanup_after_register_access(target, mstatus, number) != ERROR_OK)
+/** This function recives a program that writes a value stored at a memory
+ * address from S0 to a register, allocates the memory needed, writes the value
+ * to the memory and executes the program.
+ */
```

Comment:
Similarly to the above, I'd suggest different wording:

```suggestion
/** 
 * This function allocates 64 bits of scratch memory, fills it with the provided `value`,
 * fills the stratch address into register s0 and executes program buffer.
 * The scratch memory is deallocated at the end.
 *
 * Use case: Writing of D-extension registers (64-bit) on RV32, which needs to be done through scratch memory.
 *
 * The caller should save s0.
 * The caller should prepare a suitable program that stores the value from the memory to the register.
 */
```

Reply:
Addressed, similarly to the above.